FastAPI routes for configuration

jdavcs opened the pull request:
This is ready for review.

NOTES:
I am not at all happy with the "catch-all" nature of the configuration controller's endpoints. But I assume we need all or most of them, and for now, while we haven't converted the other controllers, maybe keeping them here is the simplest solution. I think it will be easier to reorganize endpoints once we have all the controllers done. (I kept `tool_lineages` for this reason too.)

I've added `common.py` for code shared by controllers. Since `api/__init__.py` is used for shared dependencies used by fastapi's DI system, I think a separate module is more appropriate. 

Ref: #10889.

jdavcs committed:
Add API test for 'whoami' endpoint

jdavcs committed:
Pass same formatting options to model.json()

jdavcs committed:
Add whoami fastapi endpoint

jdavcs committed:
Make tours fastapi methods async (cleanup)

jdavcs committed:
Rename controller class

jmchilton commented on lib/galaxy_test/api/test_configuration.py:
```diff
@@ -23,6 +24,11 @@
+    def test_whoami(self):
+        response = self._get("whoami")
+        self._assert_status_code_is(response, 200)
+        assert response.json()['email'] == DEFAULT_TEST_USER
```
This will pass on CI but isn't correct - maybe change this to TEST_USER not DEFAULT_TEST_USER at least.

jdavcs replied:
Fixed, thanks!

jdavcs committed:
Do not use async for blocking op; fix test

jdavcs committed:
Remove unused dependency from whoami route

jdavcs committed:
Handle api/configuration endpoint

Also, refactor config serializer: remove url_for(); I don't think this
will cause any errors.

jdavcs committed:
Remove return from finally block/do not silence error

jdavcs committed:
Keep it simple.

jdavcs committed:
Add+modify fastapi api/version endpoint

jdavcs committed:
Add unit test

jdavcs committed:
Keep method for backward compatibility

get_config_dict() is used in templates/galaxy_client_app.mako

jdavcs committed:
Add api test for api/version endpoint

jdavcs committed:
Add fastapi api/configuration/decode_id endpoint

+ Refactor some duplicate code

jdavcs committed:
Add fastapi api/configuration/dynamic_tool_confs endoint

jdavcs committed:
Fix bug, modify + add fastapi tool_lineages endpoint

Modification: do not include tools with no lineage

jdavcs committed:
Add fastapi configuration/toolbox endpoint

jdavcs committed:
Add configuration tag to fastapp

jdavcs committed:
Refactor; move logic into manager; add typing

jdavcs committed:
Add metadata to endpoints

davelopez commented on lib/galaxy/managers/users.py:
```diff
@@ -48,6 +49,17 @@
+class UserModel(BaseModel):
+    """User in a transaction context."""
+    id: int = Field(title='ID', description='User ID')
+    username: str = Field(title='username', description='User username')
```
The first letter of the title should be capitalized to display it in the docs? Otherwise, I think the title will be redundant since it will get the name of the field by default.
Would be nice to keep the consistency in the OpenAPI docs, this is something that can go to the guidelines docs as well. Maybe is a good idea to do a documentation review and formatting once all routes are documented.

jdavcs replied:
Will change. And yes - definitely +1 on documentation review, formatting all meta, docstrings, etc. for consistency.

davelopez commented on lib/galaxy/webapps/galaxy/api/configuration.py:
```diff
@@ -2,28 +2,149 @@
+    def decode_id(
+        self,
+        trans: ProvidesAppContext = Depends(get_trans),
+        *,
```
Is this `*` needed here?

jdavcs replied:
Yep. Otherwise you'll get a syntax error: non-default argument (encoded_id) follows default argument (trans). It's a common thing, I think it's mentioned somewhere in fastapi docs, but here's the PEP: https://www.python.org/dev/peps/pep-3102/#specification

mvdbeek replied:
I think we need to move the id encoding and decoding into the id fields. I don't think trans should be needed to decode an id. I started looking into that and it would be a hige simplification for all serialization/deserialization. In the meantime, is there a problem moving trans to the last argument ? That seems like it'd be a more conventional setup.

jmchilton replied:
ConfigurationManager has a self.app - that can be used regardless of what cool type fixes we can do.

jdavcs replied:
^looking into this. Moving to WIP for now.

jdavcs replied:
Replaced str with EncodedDatabaseIdField. Adding one `typing.cast` fixed the typing issue, hurray! 
Also, specifying the encoded_id as a Query object solved the "non-default arg after default arg" error.
Now looking into dropping trans in the decode endpoint.

Thank you, everyone, for the detailed reviews!

jdavcs replied:
> I think we need to move the id encoding and decoding into the id fields. I don't think trans should be needed to decode an id. I started looking into that and it would be a hige simplification for all serialization/deserialization. In the meantime, is there a problem moving trans to the last argument ? That seems like it'd be a more conventional setup.

Done in 00b7ae8 .

davelopez commented on lib/galaxy/webapps/galaxy/api/configuration.py:
```diff
@@ -2,28 +2,149 @@
+        return self.configuration_manager.tool_lineages()
+
+    @router.put(
+        '/api/configuration/toolbox',
```
```suggestion
        '/api/configuration/toolbox',
        dependencies=[AdminUserRequired],
```

jdavcs replied:
Good catch, thanks!

davelopez reviewed (commented):
Nice! :)

jdavcs committed:
Update lib/galaxy/webapps/galaxy/api/configuration.py

Co-authored-by: David López <46503462+davelopez@users.noreply.github.com>

jdavcs committed:
Capitalize first letter in model's title attr

As per code review, thanks @davelopez

jdavcs committed:
Do not rebind variable to different type

As per code review and discussion on gitter

jdavcs committed:
Use EncodedDatabaseIdField; add typing, QueryParam

Addresses code review

jdavcs committed:
Fix bug in EncodedDatabaseIdField

jdavcs committed:
Do not use transaction obj to decode an id

jdavcs committed:
Add more typing; type alias

jmchilton commented on lib/galaxy/webapps/galaxy/api/tours.py:
```diff
@@ -38,17 +38,17 @@ class FastAPITours:
 
     @router.get('/api/tours/{tour_id}')
-    def show(self, tour_id: str) -> TourDetails:
+    async def show(self, tour_id: str) -> TourDetails:
```
Doesn't this do synchronous I/O?

```python
            with open(tour_path) as f:
                tour = yaml.safe_load(f)
                load_tour_steps(tour)
```

I thought we were supposed to drop these asyncs.

jdavcs replied:
I missed this, thank you! I made the change to async for config *and* tours, *then* we had that discussion and I changed it back - but only for config! Correcting now...

jdavcs committed:
Do not use async in tours fastapi controller
